include/rmm/detail/format.hpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <array>
+#include <cstdio>
+#include <ios>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
+namespace RMM_NAMESPACE {
+namespace detail {
+
+/**
+ * @brief Format a message string with printf-style formatting
+ *
+ * This function performs printf-style formatting to avoid the need for fmt
+ * or spdlog's own templated APIs (which would require exposing spdlog
+ * symbols publicly) and returns the formatted message as a `std::string`.
+ *
+ * @param format The format string
+ * @param args The format arguments
+ * @return The formatted message
+ * @throw rmm::logic_error if an error occurs during formatting
+ */
+template <typename... Args>
+std::string formatted_log(std::string const& format, Args&&... args)
+{
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
+  auto size = std::snprintf(nullptr, 0, format.c_str(), args...) + 1;
+  RMM_EXPECTS(size >= 0, "Error during formatting.");
+  if (size == 0) { return {}; }
+  // NOLINTNEXTLINE(modernize-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+  std::unique_ptr<char[]> buf(new char[size]);
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
+  std::snprintf(buf.get(), size, format.c_str(), args...);
+  return {buf.get(), buf.get() + size - 1};  // drop '\0'
```

Comment:
Argh, sorry @harrism, I missed the +1 here when we're obtaining size. So I think a bunch of these conditions have fencepost errors.

I think we `RMM_EXPECTS(size > 0)` and then `if (size == 1) { return {}; }`

Or, I suppose we can do:

```suggestion
  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
  auto size = std::snprintf(nullptr, 0, format.c_str(), args...);
  // size is length of string without trailing '\0'
  RMM_EXPECTS(size >= 0, "Error during formatting.");
  if (size == 0) { return {}; }
  // NOLINTNEXTLINE(modernize-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
  std::unique_ptr<char[]> buf(new char[size + 1]);
  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)
  std::snprintf(buf.get(), size + 1, format.c_str(), args...);
  return {buf.get(), buf.get() + size};
```

WDYT?

Reply:
I saw this after I committed a similar but DRYer version. :)

Reply:
Yes, thanks!